Remove hash from emitter in PCDCollection

artwyman opened the pull request:
This feature is vestigial, no longer used.  App updates on PCD changes are driven by calls to update().  [Update: this isn't true, the change callback is needed by React, but the hash is not.  See further comments below.]

Repeatedly serializing all PCDs was causing slowdowns, including in applying feed actions.  Closes #1270.

On my local dev server, on an account with 800+ frogs, this change takes the `applyActions()` part of sync from ~1800ms to ~25ms.

ichub commented:
I am quite suspicious of this change. I'm not sure that it takes into account all of the reasons that this hash emitter was necessary.

> This feature is vestigial, no longer used. App updates on PCD changes are driven by calls to update().

I have not verified that this is true, and even if it was, relying on this fact to remain true seems incorrect to me for to reasons.

See `appHooks.ts`:

```
export function usePCDCollection(): PCDCollection {
  const pcds = useSelector<PCDCollection>((s) => s.pcds, []);
  return pcds;
}

export function usePCDs(): PCD[] {
  const pcds = usePCDCollection();
  return [...pcds.getAll()];
}
```

`usePCDs` is a hook, which you can roughly treat as having the following functionality: "i am a function that you can call inside of a UI component that triggers a re-render of the UI whenever the *reference* stored in the global `AppState.pcds` variable changes from one PCD collection to another.

Consequently, even if every change to the `PCDCollection` was routed through an `update({})` invocation, unless the reference changes (which I almost never happens), then all the UI components that previously used the hook that you removed would now buggy-ly *not* re-render when something inside of `PCDCollection` changes.

The hook you removed had several purposes - one of which was to be used as a sort of stable 'key' that identifies the contents of a `PCDCollection` for the purposes of sync, and the other of which which I think you have not had much if any exposure to yet was to serve as a field whose value changes to reference another string, thereby triggering a re-render of components that used the hook `usePCDCollectionWithHash`.

One way to make the performance improvement you want, while retaining the attribute the original implementation had of updates to the pcd collection triggering re-renders would be to make the emitter emit something whose reference changes that isn't necessarily some pure function (like a hash of the contents), eg. a `Wrapper<never>`.

We can discuss more in person if you'd like, I think it would be helpful to give more of an overview of how we use react in our codebase.

artwyman commented:
Thanks, that's exactly the kind of context I need.  I think I understand how state changes trigger sync, and that's independent of this PR, but how React triggers updates is still magical to me.  Let's talk about in more detail when I get to the office.  I agree that some sort of change notification without a hash should address the performance concern.  I'd also like to discuss whether there's a reasonable way a test could validate this behavior and detect when it breaks.

artwyman commented:
I've altered this PR to an alternative solution which maintains the emitter, but doesn't calculate a hash.  Instead the app hooks simply set a piece of state to a new empty object to force React to re-render.

My first thought of using a sequence number ran into a dependency warning which I think would've required a `useRef` to work around.  The new empty object seemed the simplest hack to capture the behavior I wanted.  I prefer that the fake-updating-state remain outside of the PCDCollection itself, since it's a UI-specific artifact.

This seems to average about 40ms to apply updates (I saw values ranging 20-60ms).

artwyman commented:
I did some manual testing to confirm this doesn't break reactivity:
- Manually removed some frogs and saw that they disappeared from the list.
- Added some semaphore signature PCDs via consumer-client, waited for background sync, and saw them appear on the homescreen.
- Added a test function on window which would find and delete one of those signature PCDs.  Ran that function from console, and saw a PCD disappear from the list.

@ichub let me know if there is anything else you'd like to me cover, otherwise accept when you think this is ready to merge.
